File: Python/Algorithms/1582.py
class Solution:
    def numSpecial(self, mat):
        occupied_columns = set()
        rows, columns = len(mat), len(mat[0])
        special = 0
        r_range, c_range = range(rows), range(columns)
        
        for r in r_range:
            for c in c_range:
                # found '1'
                if mat[r][c]:
                    # '1' present in a column, go to next row.
                    if c in occupied_columns:
                        break

                    bail = False
                    # reaching this point means we haven't seen
                    # another '1' so far in row, search till end.
                    row = mat[r]
                    for tc in range(c+1, columns):
                        # if we find another '1', add it to occ_cols
                        # and bail.
                        if row[tc]:
                            occupied_columns.update({c, tc})
                            bail = True
                            break
                    if bail:
                        break        
                    # Check em all here. 
                    for tr in r_range:
                        # if we find another '1', add it to occ_cols
                        # and bail. (don't count current row.)
                        if mat[tr][c] and tr != r:
                            occupied_columns.add(c)
                            bail = True
                            break
                    if bail:
                        break 

                    # add 'c' to occupied columns:
                    occupied_columns.add(c)    
                    # reaching this point means we have a special position.
                    special = special + 1
                    break
            
        # Finally, return.
        return special

File: Rust/Algorithms/1582.rs
use std::collections::HashSet;

impl Solution {
    pub fn num_special(mat: Vec<Vec<i32>>) -> i32 {
        let mut occupied_columns = HashSet::new();
        let (rows, columns) = (mat.len(), mat[0].len());
        let mut special = 0;

        for r in 0..rows {
            let mut bail = false;
            for c in 0..columns {
                if mat[r][c] == 1 {
                    if occupied_columns.contains(&c) {
                        break;
                    }

                    // Check the rest of the row:
                    for k in c + 1..columns {
                        if mat[r][k] == 1 {
                            occupied_columns.insert(k);
                            bail = true;
                            break;
                        }
                    }
                    if bail {
                        occupied_columns.insert(c);
                        break;
                    }
                    // Check all other rows:
                    for k in 0..rows {
                        if mat[k][c] == 1 && k != r {
                            occupied_columns.insert(c);
                            bail = true;
                            break;
                        }
                    }
                    if bail {
                        break;
                    }

                    // done:
                    occupied_columns.insert(c);
                    special += 1;
                }
            }
        }
        special
    }
}
